lib/spack/spack/cmd/view.py

```diff
@@ -290,14 +451,48 @@ def visitor_statlink(specs, args):
 def view(parser, args):
     'Produce a view of a set of packages.'
 
+    path = args.path[0]
+    layout = ViewYamlDirectoryLayout(spack.store.layout, path)
+
     # Process common args
-    seeds = [spack.cmd.disambiguate_spec(s) for s in args.specs]
+    if getattr(args, "all", False):
+        seeds = [spack.cmd.disambiguate_spec(s)
+                 for s in all_specs_in_view(path)]
+
+        if len(seeds) == 0:
+            tty.warn("Found no specs in %s" % path)
+    else:
+        seeds = [spack.cmd.disambiguate_spec(s) for s in args.specs]
     specs = flatten(seeds, args.dependencies.lower() in ['yes', 'true'])
     specs = filter_exclude(specs, args.exclude)
 
+    # assert that no spec is globally activated
+    for s in specs:
+        activated = filter(lambda s: s.package.is_extension
+                           and s.package.activated, specs)
+
+        if len(activated) > 0:
+            tty.error("Globally activated extensions cannot be used in "
+                      "conjunction with filesystem views. "
+                      "Please deactivate the following specs: ")
+            spack.cmd.display_specs(activated, flags=True, variants=True,
+                    long=args.verbose)
+            return
+
+    # now that we have loaded all necessary specs from the database we can
+    # monkey patch the directory layout so that all activation mechanisms will
+    # use the view based approach instead
+    setattr(spack.store, "layout", layout)
```

Comment:
I'm imagining to make this cleaner (IMO), the DirectoryViewLayout could be replaced by a ViewExtensionLayout, the point being that all current logic is located in the same place, but the object you created is responsible entirely for handling extensions (in a sense it implements an interface that only handles the extension logic). The existing YamlDirectoryLayout could also implement these functions.

* Extract ```DirectoryLayout._extension_map``` into a standalone function (I think accepting ```all_specs``` as a parameter to construct ```specs_by_hash``` would work for this).
* ```Package.activate``` now takes an ExtensionLayout (either the ViewExtensionLayout or the YamlDirectoryLayout) so spack.store.layout is always referring to the spack installation prefix for the package, and the ExtensionLayout provides the target directory to symlink into
* ViewExtensionLayout and YamlDirectoryLayout both offer functions to: retrieve an extension map for a spec; provide the target prefix to activate an extension spec in
* The ```spack activate``` command would then provide the YamlDirectoryLayout implementation of the implicit ExtensionLayout interface to ```Package.activate```

In this approach, you don't need to replace ```spack.store.layout``` and the ViewExtensionLayout wouldn't extend YamlDirectoryLayout so there wouldn't be confusion over what functions are needed from it.

Although this is a significant shift in ideas, my intent is that it is not a large change in terms of adding or removing code - not that this is necessarily the thing to optimize, but I'm saying this so if you get a different impression then I should probably explain myself better.

If this is unclear or disagreeable let me know (I'm assuming this explanation is insufficient).